### ISSUE TITLE
The server memory keeps growing during the training with multi-threading。

### ISSUE THREAD

**Issue by Lisennlp** [pytorch/xla#1512]
## 🐛 Bug

I'm using a torch-xla-nightly, which is multi-threaded, but over the course of the training, each 20-step training adds about 100 M of memory. My model  structure is  transformer. batch size is 8，It's about 13,000 steps totally, When  training,  at this rate, 65 G of memory per epoch,  over 400 G memory for six epochs, which is hard to accept.
## To Reproduce

Here's my training

    def tpu_training_loop(model, loader, device, context):
        """ Called by torch_xla_py.data_parallel. This function is executed on each core of the TPU once per epoch"""
        no_decay = ['bias', 'LayerNorm.bias', 'LayerNorm.weight']
        param_optimizer = list(model.named_parameters())
        optimizer_grouped_parameters = [
            {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay) and (
                    'bert' in n or 'decoderEmbedding' in n)],
             'weight_decay_rate': 0.01,
             'lr': args.encoder_learning_rate},
            {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay) and (
                    'bert' not in n and 'decoderEmbedding' not in n)],
             'weight_decay_rate': 0.01,
             'lr': args.decoder_learning_rate},
            {'params': [p for n, p in param_optimizer if
                        any(nd in n for nd in no_decay) and (
                                'bert' in n or 'decoderEmbedding' in n)],
             'weight_decay_rate': 0.0,
             'lr': args.encoder_learning_rate},
            {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay) and (
                    'bert' not in n and 'decoderEmbedding' not in n)],
             'weight_decay_rate': 0.0,
             'lr': args.decoder_learning_rate}, ]

        optimizer = context.getattr_or(
            'optimizer',
            BERTAdam(optimizer_grouped_parameters, lr=args.encoder_learning_rate,
                     warmup=args.warmup_proportion,
                     t_total=num_train_steps))

        if args.warmup_proportion is not None:
            warmup_steps = int(args.warmup_proportion * num_train_steps + 0.5)

        optimizer.zero_grad()
        tr_loss = None
        pbar = None
        if str(pbar_device) == str(device):  # All threads are in sync. Use progress bar only on one of them
            pbar = tqdm(total=int(pbar_steps), desc=f"training ", dynamic_ncols=True)
        tracker = tpu_xm.RateTracker()
        model.train()
        for step, batch in enumerate(loader):
            input_ids, input_mask, segment_ids, position_ids, answer_ids, answer_mask = batch
            log_prob, _ = model(input_ids,
                                segment_ids,
                                input_mask,
                                position_ids,
                                answer_ids=answer_ids,
                                answer_mask=answer_mask)
            loss = criterion(log_prob, answer_ids[:, 1:].contiguous().view(-1))
            if args.gradient_accumulation_steps > 1:
                loss = loss / args.gradient_accumulation_steps
            loss.backward()
            tracker.add(args.train_batch_size)
            tr_loss = loss * args.gradient_accumulation_steps if step == 0 else tr_loss + loss * args.gradient_accumulation_steps
            if pbar is not None:
                pbar.update(1)
            if (step + 1) % args.gradient_accumulation_steps == 0:
                tpu_xm.optimizer_step(optimizer)
                optimizer.zero_grad()
        return tr_loss.item() / step

## Environment
torch==1.0.1post2
python==3.6.9

 - reproducible on XLA backend [CPU/TPU]:
 - torch_xla version: torch-xla-nightly
 - Any other relevant information. In case of issues related to slow executions, re-compilations, etc.., it is useful to have the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

**Comment by zcain117** [comment]
Could you collect the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information ?

This helps us track down which part of the model keeps growing

**Comment by dlibenzi** [comment]
It is hard to debug from that incomplete snippet.
To see if it makes any difference, can you try with accumulation step 1, and using SGD?

**Comment by Lisennlp** [comment]
@dlibenzi I tried the SGD optimizer, but got the same result.  GCE memory is still increasing with training.However,  I don't have that problem with GPU.I really do not know what to do.

**Comment by Lisennlp** [comment]
@zcain117 Thanks. I'll give it a shot.

**Comment by dlibenzi** [comment]
Try to use the sample generator as source, to see if problems are still there:

https://github.com/pytorch/xla/blob/d62a57564978bdb4af56317aed454fdf0b2b62ff/test/test_train_mp_mnist.py#L61

**Comment by Lisennlp** [comment]
@dlibenzi  Thank for your reply！I used this method in [xla/test/test_train_mp_mnist.py to construct the fake data, and the GCE memory kept growing as well](https://github.com/pytorch/xla/blob/d62a57564978bdb4af56317aed454fdf0b2b62ff/test/test_train_mp_mnist.py#L61)

**Comment by zcain117** [comment]
This (and the other issue you're seeing) might be similar to https://github.com/pytorch/xla/issues/1379 where the graph kept growing. To help us determine, it's useful to collect the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

**Comment by dlibenzi** [comment]
Since multi-threading is deprecated, and we have the other issue open, I will close this one.
